### PR TITLE
Support hack-style pipeline

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Supports [hack-style pipeline](https://github.com/js-choi/proposal-hack-pipes). Also deprecates [smart-style pipeline](https://github.com/js-choi/proposal-smart-pipelines).

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/parse/babel.js [hunk 1]
```diff
@@ -39,8 +38,9 @@ const parseOptions = {
   ranges: true,
 };
 const recordAndTuplePlugin = ["recordAndTuple", { syntaxType: "hash" }];
+const v8intrinsicPlugin = "v8intrinsic";
 const pipelineOperatorPlugins = [
-  ["pipelineOperator", { proposal: "smart" }],
+  ["pipelineOperator", { proposal: "hack", topicToken: "%" }],
```
Comment (sosukesuzuki) [inline src/language-js/parse/babel.js:43]: `%` is used in official explainer ( https://github.com/js-choi/proposal-hack-pipes ). So we follow it.

File: src/language-js/parse/babel.js [hunk 2]
```diff
@@ -111,11 +111,20 @@ function createParse(parseMethod, ...optionsCombinations) {
     }
 
     if (text.includes("|>")) {
-      combinations = pipelineOperatorPlugins.flatMap((pipelineOperatorPlugin) =>
+      const conflictsPlugins = pipelineOperatorPlugins;
+      if (text.includes("%")) {
+        // @ts-expect-error
+        conflictsPlugins.push(v8intrinsicPlugin);
```
Comment (sosukesuzuki) [inline src/language-js/parse/babel.js:117]: Now, we cannot combine v8intrinsic plugin and hack-style pipeline.

File: src/language-js/parse/babel.js [hunk 3]
```diff
@@ -111,11 +111,20 @@ function createParse(parseMethod, ...optionsCombinations) {
     }
 
     if (text.includes("|>")) {
-      combinations = pipelineOperatorPlugins.flatMap((pipelineOperatorPlugin) =>
+      const conflictsPlugins = pipelineOperatorPlugins;
+      if (text.includes("%")) {
+        // @ts-expect-error
```
Comment (sosukesuzuki) [inline src/language-js/parse/babel.js:116]: This will be fix by #11316

File: src/language-js/parse/babel.js [hunk 4]
```diff
@@ -111,11 +111,20 @@ function createParse(parseMethod, ...optionsCombinations) {
     }
 
     if (text.includes("|>")) {
-      combinations = pipelineOperatorPlugins.flatMap((pipelineOperatorPlugin) =>
+      const conflictsPlugins = pipelineOperatorPlugins;
```
Comment (fisker) [inline src/language-js/parse/babel.js:114]: Need copy array here.

File: src/language-js/parse/babel.js [hunk 5]
```diff
@@ -111,11 +111,20 @@ function createParse(parseMethod, ...optionsCombinations) {
...
+      if (text.includes("%")) {
```
Comment (fisker) [inline src/language-js/parse/babel.js:115]: I think V8 intrinsics alway start with a uppercase letter, so `/%[A-Z]/.test`?

**Comment by sosukesuzuki** [comment]
@fisker Thank you for reviewing. I've addressed your comments.

File: src/language-js/parse/babel.js (hunk 4 above)
Comment (fisker) [reply src/language-js/parse/babel.js]: This is not fixed, you are still manipulating to the `pipelineOperatorPlugins` array.

File: src/language-js/parse/babel.js [hunk 6]
```diff
@@ -121,12 +123,20 @@ function createParse(parseMethod, ...optionsCombinations) {
       );
     }
 
+    const shouldEnableV8intrinsicPlugin = /%[A-Z]/.test(text);
     if (text.includes("|>")) {
+      if (shouldEnableV8intrinsicPlugin) {
+        pipelineOperatorPlugins.push(v8intrinsicPlugin);
```
Comment (fisker) [inline src/language-js/parse/babel.js:129]: This line not covered.

File: src/language-js/parse/babel.js (hunk 4 above)
Comment (sosukesuzuki) [reply src/language-js/parse/babel.js]: Ah sorry I misunderstood. Okay, I'll fix it.

File: src/language-js/parse/babel.js [hunk 7]
```diff
@@ -125,10 +125,11 @@ function createParse(parseMethod, ...optionsCombinations) {
 
     const shouldEnableV8intrinsicPlugin = /%[A-Z]/.test(text);
     if (text.includes("|>")) {
+      const conflictsPlugins = [...pipelineOperatorPlugins];
       if (shouldEnableV8intrinsicPlugin) {
-        pipelineOperatorPlugins.push(v8intrinsicPlugin);
+        conflictsPlugins.push(v8intrinsicPlugin);
       }
```
Comment (fisker) [inline src/language-js/parse/babel.js]:
```suggestion
      const conflictsPlugins = shouldEnableV8intrinsicPlugin
        ? [...pipelineOperatorPlugins, v8intrinsicPlugin]
        : pipelineOperatorPlugins;
```

**Comment by fisker** [comment]
Let's add a failed test too

**Prettier pr-11335**
[Playground link](https://deploy-preview-11335--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEGCGMCuamAvJgKQDicMA8gA4wCWAtowF56PQDKMehACgBmUAJSYAPgD5MkKGggAbOADpFEAOYCyogNwgANCAgMu85KBwAnKxADuABWsI0yEDkV2cAT1dGARlY4YADW1Nx0wYxQGsgwVvhwRnDM-nAAJukZADI4Mfg4GnAAYhBWzHhMMW44+DAQhiAAFjDMigDqTYzwaJFgcNwu3YwAbt3ebmBofiDRaHBWMA5BGhXIQh7zRgBWaAAeAEJBoeE4zHDZ0XDrm0kgu3vc0RrKAIr4EPA3ilsgkVbzKxufw4NKKRp0KzRGDtRjpGBNZAADgADEZIRB5u0gnQ3JC4ICRtcjABHD7wZYmVwoHBoAC0UDgGQyjSscDJjDZy0KayQGx+d3mrDiCUFzze5OufNuRj4-lh8MRSAATLKgoxFM8AMIQZi8kAEgCsjUIcAAKqDqfzfiNEgBJKBZWDcMBQhgAQUdvG8ym+8wAvv6gA)

```sh
--parser babel
```

**Input:**

```jsx
const status = %GetOptimizationStatus(fn) |> console.log(%);
```

**Output:**

```jsx
SyntaxError: Topic reference is unbound; it must be inside a pipe body. (1:16)
> 1 | const status = %GetOptimizationStatus(fn) |> console.log(%);
    |                ^
```

File: src/language-js/parse/babel.js [hunk 8]
```diff
@@ -121,12 +123,21 @@ function createParse(parseMethod, ...optionsCombinations) {
       );
     }
 
+    const shouldEnableV8intrinsicPlugin = /%[A-Z]/.test(text);
     if (text.includes("|>")) {
-      combinations = pipelineOperatorPlugins.flatMap((pipelineOperatorPlugin) =>
+      const conflictsPlugins = [...pipelineOperatorPlugins];
+      if (shouldEnableV8intrinsicPlugin) {
+        conflictsPlugins.push(v8intrinsicPlugin);
+      }
```
Comment (fisker) [inline src/language-js/parse/babel.js:131]:
```suggestion
      const conflictsPlugins = shouldEnableV8intrinsicPlugin
        ? [...pipelineOperatorPlugins, v8intrinsicPlugin]
        : pipelineOperatorPlugins;
```

If we don't need add  `v8intrinsicPlugin`, there is no need copy that array.

**Comment by fisker** [comment]
I believe this is just basic support, because `%` can be anywhere, many node check logic didn't consider it.

But we can improve in future, I guess.

**Review by fisker — APPROVED** [review]

**Pull request merged**